### PR TITLE
Acrn dev virtio gpu upstream

### PR DESCRIPTION
**Pull request by psun3x**
https://github.com/projectacrn/acrn-hypervisor/issues/7210

### CODE REVIEW

**Review by ywan170 — APPROVED** [review]

**Comment by acrnsi** [comment]
Can one of the admins verify this patch?

**Comment by psun3x** [comment]
The build system dependency library: libpixman-1-dev,   libsdl2-dev, libegl-dev, libgles-dev

**Comment by yakuizhao** [comment]
This patch still has some problems and needs to be updated.
Thanks

**Comment by ywan170** [comment]
ywan170@ywan170-OptiPlex-7050:~/work2/acrn-hypervisor/devicemodel$ git am ~/incoming/p*
Applying: dm: virtio-gpu: basic framework
Applying: dm: virtio-gpu: virtual display support
.git/rebase-apply/patch:569: space before tab in indent.
        }
warning: 1 line adds whitespace errors.
Applying: dm: virtio-gpu: EDID emulation of virtual monitor
Applying: dm: virtio-gpu: 2D mode support
Applying: dm: virtio-gpu: cursor surpport
.git/rebase-apply/patch:280: space before tab in indent.
        }
warning: 1 line adds whitespace errors.
Applying: dm: virtio-gpu: VGA compability support

Please fix all warnings.

**Review by gvancuts — CHANGES_REQUESTED** [review]

**Comment by gvancuts** [comment]
> The build system dependency library: libpixman-1-dev, libsdl2-dev, libegl-dev, libgles-dev

Is the glib dev package pulled in automatically via any of those?

Remember to update the documentation *and* the Debian package dependencies (both build and runtime).

File: devicemodel/hw/gc.c [hunk 1]
```diff
@@ -0,0 +1,84 @@
+#include <stdlib.h>
```
[the remark this answers is not in the document]
Comment (psun3x) [reply devicemodel/hw/gc.c:1]: Will fix this in next version.

File: devicemodel/hw/pci/core.c [hunk 2]
```diff
@@ -351,6 +353,13 @@ pci_parse_slot(char *opt)
 			vsbl_set_bdf(bnum, snum, fnum);
 		}
 	}
+
+	if ((strcmp("virtio-gpu", emul) == 0)) {
+		pr_info("%s: virtio-gpu device found, activate virtual display.\n",
```
[the remark this answers is not in the document]
Comment (psun3x) [reply devicemodel/hw/pci/core.c:358]: Will fix this in next version.

File: devicemodel/hw/pci/virtio/virtio_gpu.c [hunk 3]
```diff
@@ -0,0 +1,1376 @@
+/*
+ * Copyright (C) 2021 Intel Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (psun3x) [reply devicemodel/hw/pci/virtio/virtio_gpu.c:2]: Will fix this in next version.

File: devicemodel/hw/pci/virtio/virtio_gpu.c [hunk 4]
```diff
@@ -0,0 +1,1376 @@
...
+	virtio_gpu_set_status,		/* called on guest set status */
+};
+static int virtio_gpu_device_cnt = 0;
+
+static void
+virtio_gpu_set_status(void *vdev, uint64_t status)
+{
+	struct virtio_gpu *gpu;
+
+	pr_dbg("virtio-gpu set deivce status 0x%x.\n", status);
```
[the remark this answers is not in the document]
Comment (psun3x) [reply devicemodel/hw/pci/virtio/virtio_gpu.c:342]: Will fix this in next version.

File: devicemodel/hw/pci/virtio/virtio_gpu.c [hunk 5]
```diff
@@ -0,0 +1,1376 @@
...
+	gpu->base.status = status;
+}
+
+static void
+virtio_gpu_reset(void *vdev)
+{
+	struct virtio_gpu *gpu;
+	struct virtio_gpu_resource_2d *r2d;
+
+	pr_dbg("virtio-gpu device reset.\n");
```
[the remark this answers is not in the document]
Comment (psun3x) [reply devicemodel/hw/pci/virtio/virtio_gpu.c:353]: Will fix this in next version.

File: devicemodel/hw/vga.c [hunk 6]
```diff
@@ -0,0 +1,1450 @@
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause-FreeBSD
+ *
+ * Copyright (c) 2015 Tycho Nightingale <tycho.nightingale@pluribusnetworks.com>
```
[the remark this answers is not in the document]
Comment (psun3x) [reply devicemodel/hw/vga.c:4]: Yes, I add some APIs in this file. I will add Intel copyright info in next version.

File: devicemodel/hw/vga.c [hunk 7]
```diff
@@ -0,0 +1,1450 @@
...
+			*val = vd->vga_crtc.crtc_end_vert_blank;
+			break;
+		case CRTC_MODE_CONTROL:
+			*val = vd->vga_crtc.crtc_mode_ctrl;
+			break;
+		case CRTC_LINE_COMPARE:
+			*val = vd->vga_crtc.crtc_line_compare;
+			break;
+		default:
+			pr_err("XXX VGA CRTC: inb 0x%04x at index %d\n", port, vd->vga_crtc.crtc_index);
```
[the remark this answers is not in the document]
Comment (psun3x) [reply devicemodel/hw/vga.c:802]: It is just a symbol, will remove this in next version.

File: devicemodel/include/vdisplay.h [hunk 8]
```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2021 Intel Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (psun3x) [reply devicemodel/include/vdisplay.h:2]: Will modify to "2022 Intel Corporation." in next version.

File: devicemodel/include/vdisplay.h [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+ *
```
[the remark this answers is not in the document]
Comment (psun3x) [reply devicemodel/include/vdisplay.h:3]: Will add license info in next version.

File: devicemodel/include/vga.h [hunk 10]
```diff
@@ -0,0 +1,221 @@
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause-FreeBSD
+ *
+ * Copyright (c) 2015 Tycho Nightingale <tycho.nightingale@pluribusnetworks.com>
```
[the remark this answers is not in the document]
Comment (psun3x) [reply devicemodel/include/vga.h:4]: Will remove this in next version.

**Comment by psun3x** [comment]
glib will not be pulled in automatically. We will add these dependency library in doc/getting-started/getting-started.rst

File: devicemodel/Makefile [hunk 11]
```diff
@@ -39,6 +39,10 @@ CFLAGS += -I$(BASEDIR)/include
 CFLAGS += -I$(BASEDIR)/include/public
 CFLAGS += -I$(DM_OBJDIR)/include
 CFLAGS += -I$(TOOLS_OUT)/services
+CFLAGS += -I/usr/include/pixman-1
+CFLAGS += -I/usr/include/glib-2.0
+CFLAGS += -I/usr/include/SDL2
+CFLAGS += -I/usr/include/EGL
```
Comment (saininav) [inline devicemodel/Makefile:45]: It should support cross compilation. I would prefer:
-CFLAGS += -I/usr/include/pixman-1
-CFLAGS += -I/usr/include/glib-2.0
-CFLAGS += -I/usr/include/SDL2
-CFLAGS += -I/usr/include/EGL
+CFLAGS += -I$(SYSROOT)/usr/include/pixman-1
+CFLAGS += -I$(SYSROOT)/usr/include/glib-2.0
+CFLAGS += -I$(SYSROOT)/usr/include/SDL2
+CFLAGS += -I$(SYSROOT)/usr/include/EGL

**Review by NanlinXie — APPROVED** [review]

File: devicemodel/hw/vga.c [hunk 12]
```diff
@@ -0,0 +1,1428 @@
+/*
+ * Copyright (C) 2022 Intel Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-FreeBSD
```
Comment (gvancuts) [inline devicemodel/hw/vga.c:4]: This license identifier is deprecated (see https://spdx.org/licenses/BSD-2-Clause-FreeBSD.html). Please use `BSD-2-Clause-Views` instead. See https://spdx.org/licenses/BSD-2-Clause-Views.html for more details.

File: devicemodel/hw/vga.c [hunk 13]
```diff
@@ -0,0 +1,1428 @@
...
+			*val = vd->vga_crtc.crtc_end_vert_blank;
+			break;
+		case CRTC_MODE_CONTROL:
+			*val = vd->vga_crtc.crtc_mode_ctrl;
+			break;
+		case CRTC_LINE_COMPARE:
+			*val = vd->vga_crtc.crtc_line_compare;
+			break;
+		default:
+			pr_err("VGA CRTC: inb 0x%04x at index %d\n", port, vd->vga_crtc.crtc_index);
```
Comment (gvancuts) [inline devicemodel/hw/vga.c:780]: What does `inb` mean?

File: devicemodel/hw/vga.c [hunk 14]
```diff
@@ -0,0 +1,1428 @@
...
+			   !(vga->vberegs.enable & VBE_DISPI_DISABLED)) {
+			vga->vberegs.enable = VBE_DISPI_DISABLED;
+			vga->enable = false;
+		}
+	}
+
+	if (!vga->gc->gc_image->vgamode &&
+	    vga->vberegs.xres == 0 &&
+	    vga->vberegs.yres == 0) {
+		pr_dbg("switching to VGA mode\r\n");
```
Comment (gvancuts) [inline devicemodel/hw/vga.c:1394]: `switching` -> `Switching`

Why the extra `\r` character here?

File: devicemodel/hw/vga.c [hunk 15]
```diff
@@ -0,0 +1,1428 @@
...
+		vga->gc->gc_image->vgamode = 1;
+		vga->gc->gc_image->width = 0;
+		vga->gc->gc_image->height = 0;
+	} else if (vga->gc->gc_image->vgamode &&
+		   vga->vberegs.xres != 0 &&
+		   vga->vberegs.yres != 0) {
+		pr_dbg("switching to VESA mode\r\n");
```
Comment (gvancuts) [inline devicemodel/hw/vga.c:1401]: Same comment as before

File: devicemodel/include/vga.h [hunk 16]
```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (C) 2022 Intel Corporation. All rights reserved.
```
Comment (gvancuts) [inline devicemodel/include/vga.h:2]: Remove the `All rights reserved` string

File: devicemodel/include/vga.h [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-FreeBSD
```
Comment (gvancuts) [inline devicemodel/include/vga.h:4]: Same comment about the license tag, this one is deprecated.

**Comment by NanlinXie** [comment]
@dbkinder and @amyreye , could you help to publish the GSG document since the dependency library has updated.

**Comment by gvancuts** [comment]
@NanlinXie , do we have someone who's going to finish off the documentation piece **and** fix the license and copyright header issues?

**Comment by NanlinXie** [comment]
@psun3x will update document later, doc impact analysis updated in https://github.com/projectacrn/acrn-hypervisor/issues/7210

**Comment by gvancuts** [comment]
> @psun3x will update document later, doc impact analysis updated in #7210

Will he take care of the copyright and license headers too?

File: devicemodel/include/vga.h (hunk 16 above)
Comment (yaomichael) [reply devicemodel/include/vga.h:2]: the All rights reserved string is not a must, but it does no harm. So my recommendation is: 
* if it's already there, don't bother to remove it. 
* If it's a new file, don't add it. 
In some cases, we need to keep it though. for example, we copied a snippet from https://github.com/oasis-tcs/virtio-spec/blob/master/virtio-gpu.tex, which requires we keep its copyright line including "all rights reserved".

File: devicemodel/include/vga.h (hunk 17 above)
Comment (yaomichael) [reply devicemodel/include/vga.h:4]: I feel we don't have to update a deprecated license to a newer one, especially when it was from 3rd party we leveraged. New projects should not use deprecated license tag, but the license of the older projects should be still effective.

Comment (gvancuts) [reply devicemodel/include/vga.h:4]: Thanks for your input @yaomichael , should we have kept the original copyright notice in there too then (alongside the Intel copyright). Where was this leveraged from?

File: devicemodel/include/vga.h (hunk 16 above)
Comment (gvancuts) [reply devicemodel/include/vga.h:2]: I don't see this `BSD-2-Clause-FreeBSD` license in the link you provided above. Where are those snippets coming from?

File: devicemodel/include/vga.h (hunk 17 above)
Comment (yaomichael) [reply devicemodel/include/vga.h:4]: good point... just checked and vga.h was from FreeBSD/bhyve. We need to add original copyright notice back

File: devicemodel/include/vga.h (hunk 16 above)
Comment (gvancuts) [reply devicemodel/include/vga.h:2]: Hi @yaomichael ,

I sought guidance from the legal team and here is their feedback.

> the All rights reserved string is not a must, but it does no harm. So my recommendation is:
> 
>     * if it's already there, don't bother to remove it.

They recommend we actively remove it (from the Intel copyright line only of course).

> 
>     * If it's a new file, don't add it.

Yes

>       In some cases, we need to keep it though. for example, we copied a snippet from https://github.com/oasis-tcs/virtio-spec/blob/master/virtio-gpu.tex, which requires we keep its copyright line including "all rights reserved".

True for all other, non-Intel copyright lines.

File: devicemodel/include/vga.h (hunk 17 above)
Comment (gvancuts) [reply devicemodel/include/vga.h:4]: > I feel we don't have to update a deprecated license to a newer one, especially when it was from 3rd party we leveraged. New projects should not use deprecated license tag, but the license of the older projects should be still effective.

Hi @yaomichael , just as FYI, this is indeed in line with the legal team's guidance. Thanks for your input on this!

**Pull request merged**